design/func-version.md

```diff
@@ -0,0 +1,378 @@
+responsible backwards compatible manner can lead to costly operational overhead
+for Kanister users. In addition, it also becomes difficult for Kanister
+maintainers to assess the blast radius of the change, determine the rollout
+timing, provide timely support etc.
+
+This proposal outlines a version management scheme that can be used to manage
+the lifecycle of Kanister Functions as they evolve from inception to retirement.
+
+## Proposed Solution
+
+### Summary
+
+The proposed solution uses a version management scheme to communicate the
+stability, availability, promotion and deprecation of Kanister Functions.
+
+The design extends the existing implementation of the action-level
+[`PreferredVersion` mechanism][1] to the phase-level by enabling users to
+specify the desired  Kanister Function version within the blueprint. It also
+proposes changes to the relevant interfaces with code examples demonstrating the
+implementation of multiple versions of a Kanister Function.
+
+### Assumptions And Constraints
+
+* The proposed solution is compatible with existing downstream custom Kanister
+Function implementation, without any code change.
+
+### Function Versioning
+
+The proposed solution introduces a new Go interface named `VersionedFunc`. The
+`VersionedFunc` type embeds the `Func` type to include information about the
+multiple versions of implementation of the underlying Kanister Function.
+
+```go
+// VersionedFunc extends the Func type to include information about the multiple
+// versions of the underlying Kanister Function.
+type VersionedFunc interface {
+  // Func is the v0.0.0 version of the function. For existing functions, this
+  // maps to their existing `Func` implementation.
+  Func
+
+  // StableVersion returns the stable version of the function.
+  StableVersion() string
+
+  // Versions maps the different versions of the function to their respective
+  // implementations.
+  Versions() map[string]Func
+}
+```
+
+By embedding the existing `Func` type in the new `VersionedFunc` type, existing
+Kanister Functions will continue to work as-is. Functions maintainers have the
+flexibility to decide when/if they need to extend existing Function from being a
+`Func` type to a `VersionedFunc` type.
+
+As an example, if new versions of the `KubeExec` Function are introduced, the
+existing [`kubeExecFunc` struct][2] will be updated to implement the
+`VersionedFunc` interface:
+
+```go
+func init() {
+  _ = kanister.RegisterVersionedFunc(&kubeExecFunc{})
+}
+
+var  _ kanister.VersionedFunc = (*kubeExecFunc)(nil)
+
+func (*kubeExecFunc) StableVersion() string {
+  return "v0.1.0"
+}
+
+func (*kubeExecFunc) Versions() map[string]kanister.Func {
+  return map[string]kanister.Func{
+    kanister.DefaultVersion: &kubeExecFunc{},
+    "v0.1.0": &kubeExecFuncV010{
+      parent: &kubeExecFunc{},
+    },
+    "v0.2.0": &kubeExecFuncV020{
+      parent: &kubeExecFunc{},
+    },
+  }
+}
```

Comment:
It looks like this implementation depends on the values that are there in the map where we register functions. That is the reason I was thinking if having the method as behaviour of functions would be a good idea.
Also, if my understanding is right, should we not read the values from that map, instead of just had coding here.

Reply:
or maybe I am not able to understand the exact use case of this method.

Reply:
If possible, I don't want the new caller code to directly depend on the global map going forward. Sorry for the confusion - I need to fix up this part. I imagine the actual implementation of `StableVersion()` and `Versions()` would involve returning the versions information read from the in-memory config file discussed below.

Reply:
So during `init()`, the [global registration function](https://github.com/kanisterio/kanister/pull/1639/files#diff-8a18494585b5ed63b67bb4b0ed2e7cee05065091750638787ad5bd7b57213fe6R160-R169) asks a Kanister Function for all its versions, via the `Versions()` method.

The `Versions()` method gets the versions info from the [`embed`ded](https://pkg.go.dev/embed) in-memory config file, which looks something like:

```yaml
functions:
  kubeExec:
    versions:
      served:
      - 0.1.0
      - 0.2.0
      stable: 0.1.0
      deprecated:
      - 0.0.0
  kubeTask:
    versions:
      served:
      - 0.0.0
      - 0.1.0
      - 0.1.1
      - 0.1.2
      stable: 0.1.2
```

During phase execution, we will follow the [existing version detection logic](https://github.com/kanisterio/kanister/blob/8b5eb2026220d4ef9c1c6d0ba07c395acec2c784/pkg/phase.go#L123-L148) to find the requested version from the `funcs` global map. The `StableVersion()` function let us to deduce the default version, in case the blueprint doesn't specify a function version. We can also add a `IsDeprecated(version string) bool` string to the `kanister.VersionedFunc` interface to determine if a requested version is deprecated.

WDYT?

Reply:
ok, so IIUC we don't want to access the global map that is the only reason we are trying to persist the supported versions in this file?
Also, we are slightly changing (`RegisterVersionedFunc`) how new function are going to register themselves? Because until now functions used to register themselves in their respective source files.
I am not sure if we are going to get into problems if we keep using that global map but it looks like we now have two places where we maintain the function versions, that global map and this file.

Reply:
We use this file to populate the global map. The phase execution code will still look up the global map for the function and version that the phase wants. Ideally, I want to get rid of the global map, but that will break many things.

Whatever approach we take, there will be two places with the function versions anyway. We need to load them from a source into the global map. Do we rather hard-code all the versions of all functions?

The main reason why I want to store these versions in a file is so that we can look at the repo and easily determine the states of all the function versions, at each commit. I added this info in the "Versions Discovery" section.